cms/djangoapps/contentstore/models.py

```diff
@@ -197,70 +261,90 @@ def update_or_create(
+        /,
+        upstream_container_key: LibraryContainerLocator,
+        upstream_context_key: str,
+        downstream_usage_key: UsageKey,
+        downstream_context_key: CourseKey,
+        version_synced: int,
+        version_declined: int | None = None,
+        created: datetime | None = None,
+    ) -> "ContainerLink":
         """
-        Returns a summary of links by upstream context for given downstream_context_key.
-        Example:
-        [
-            {
-                "upstream_context_title": "CS problems 3",
-                "upstream_context_key": "lib:OpenedX:CSPROB3",
-                "ready_to_sync_count": 11,
-                "total_count": 14
-            },
-            {
-                "upstream_context_title": "CS problems 2",
-                "upstream_context_key": "lib:OpenedX:CSPROB2",
-                "ready_to_sync_count": 15,
-                "total_count": 24
-            },
-        ]
+        Update or create entity link. This will only update `updated` field if something has changed.
         """
-        result = cls.filter_links(downstream_context_key=downstream_context_key).values(
-            "upstream_context_key",
-            upstream_context_title=F("upstream_block__learning_package__title"),
-        ).annotate(
-            ready_to_sync_count=Count("id", Q(ready_to_sync=True)),
-            total_count=Count('id')
-        )
-        return result
+        if not created:
+            created = datetime.now(tz=timezone.utc)
+        new_values = {
+            'upstream_container_key': upstream_container_key,
+            'upstream_context_key': upstream_context_key,
+            'downstream_usage_key': downstream_usage_key,
+            'downstream_context_key': downstream_context_key,
+            'version_synced': version_synced,
+            'version_declined': version_declined,
+        }
+        if upstream_container:
+            new_values.update(
+                {
+                    'upstream_container': upstream_container,
+                }
+            )
+        try:
+            link = cls.objects.get(downstream_usage_key=downstream_usage_key)
+            # TODO: until we save modified datetime for course xblocks in index, the modified time for links are updated
+            # everytime a downstream/course block is updated. This allows us to order links[1] based on recently
+            # modified downstream version.
+            # pylint: disable=line-too-long
+            # 1. https://github.com/open-craft/frontend-app-course-authoring/blob/0443d88824095f6f65a3a64b77244af590d4edff/src/course-libraries/ReviewTabContent.tsx#L222-L233
+            has_changes = True  # change to false once above condition is met.
+            for key, value in new_values.items():
+                prev = getattr(link, key)
+                # None != None is True, so we need to check for it specially
+                if prev != value and ~(prev is None and value is None):
```

Comment:
Is this actually necessary here? I know it's a concern when we're querying SQL, but these comparisons are at the Python layer, right?

Reply:
Yeah, something odd here. This part is also unchanged from master, so would it be OK if @navinkarkera opens a separate PR to clean it up? I'm not sure of the details so not confident refactoring it at the last minute.

Reply:
That's fine. But am I wrong that `has_changes` can never be False here? That seems like a straight-up bug?

Reply:
@bradenmacdonald Yes, until we store modified time in course xblocks, this was hardcoded to be True.

Reply:
>Is this actually necessary here? I know it's a concern when we're querying SQL, but these comparisons are at the Python layer, right?
> Yeah, something odd here. None != None is True is not accurate in python.

Now, I am confused too. I remember facing with null values while comparing here but cannot remember the exact issue. I'll remove this additional check for now.